src/partr.c

```diff
@@ -228,6 +219,170 @@ static int multiq_check_empty(void)
 }
 
 
+static inline void multiq_init(void)
+{
+    heap_p = heap_c * jl_n_threads;
+    heaps = (taskheap_t *)calloc(heap_p, sizeof(taskheap_t));
+    for (int32_t i = 0; i < heap_p; ++i) {
+        uv_mutex_init(&heaps[i].lock);
+        heaps[i].tasks = (jl_task_t **)calloc(tasks_per_heap, sizeof(jl_task_t*));
+        jl_atomic_store_relaxed(&heaps[i].ntasks, 0);
+        jl_atomic_store_relaxed(&heaps[i].prio, INT16_MAX);
+    }
+    unbias_cong(heap_p, &cong_unbias);
+    jl_gc_mark_enqueued_tasks = &multiq_gc_mark_enqueued_tasks;
+    partr_enqueue_task = &multiq_insert;
+    partr_dequeue_task = &multiq_deletemin;
+    partr_check_empty = &multiq_check_empty;
+}
+
+
+
+// work-stealing deque
+
+// The work-stealing deque by Chase and Lev (2005).  Le et al. (2013) provides
+// C11-complienet memory ordering.
+//
+// Ref:
+// * Chase and Lev (2005) https://doi.org/10.1145/1073970.1073974
+// * Le et al. (2013) https://doi.org/10.1145/2442516.2442524
+//
+// TODO: Dynamic buffer resizing.
+typedef struct _wsdeque_t {
+    union {
+        struct {
+            jl_task_t **tasks;
+            _Atomic(int64_t) top;
+            _Atomic(int64_t) bottom;
+        };
+        uint8_t padding[JL_CACHE_BYTE_ALIGNMENT];
+    };
+} wsdeque_t;
+
+static wsdeque_t *wsdeques;
+
+
+static int wsdeque_push(jl_task_t *task, int16_t priority_ignord)
+{
+    int16_t tid = jl_threadid();
+    int64_t b = jl_atomic_load_relaxed(&wsdeques[tid].bottom);
+    int64_t t = jl_atomic_load_acquire(&wsdeques[tid].top);
```

Comment:
What is being load-acquired? There are no stores in `wsdeque_steal`
```suggestion
    int64_t t = jl_atomic_load(&wsdeques[tid].top); // ensures that `tasks` is successfully stolen before we try to reuse the slot below
```

Reply:
I'm trying to digest this and I realized that I still don't get it. As we discussed, there is a CAS on `.top` in `wsdeque_steal`

https://github.com/JuliaLang/julia/blob/0e50d013553f0ff54f41ac305be57cf29660cb99/src/partr.c#L316-L318

The loaded task only matters in the success path where we have a seq_cst write that supersets release write. So, we have:

* a sequenced-before edge from the load `.tasks[t % tasks_per_heap]` to the store on the `.top` via CAS in the next line (as they both happens in the same thread)
* a synchronizes-with edge from the store on the `.top` to the load of `.top` in `wsdeque_push` you quoted above

So they establish happens-before and it looks like we know that the task is loaded by the time we load the `.top`? IIRC, your concern was that loading `.tasks[t % tasks_per_heap]` and storing `.top` can be reordered, but doesn't release forbids that?

> A store operation with this memory order performs the release operation: no **reads or writes** in the current thread can be reordered after this store.
> https://en.cppreference.com/w/cpp/atomic/memory_order

Reply:
I am not sure. But edges are not transitive, unless all of them are seq-cst, IIUC.

Reply:
But yes, looks like we have a proper release/acquire pair here on top to ensure the ops on tasks are okay.

Reply:
Hmm.... Lahav et al.'s definition of happens-before `hb` is a transitive closure of the union of sequenced-before `sb` and synchronizes-with `sw`

![image](https://user-images.githubusercontent.com/29282/151283975-d6479a6e-6148-4130-abf9-148a672a05f2.png)

`R^+` is the transitive closure of `R`:

![image](https://user-images.githubusercontent.com/29282/151283785-2cd48af5-cb6a-44f0-9ec5-a13eec70d775.png)

But I don't know if that's a mismatch between C11 semantics (though they use the same notions for discussing C11 so probably not) and/or some difference to the actual definition in the standard.